### ISSUE TITLE
Option of List generates invalid codec

### ISSUE THREAD

**Issue by alltonp** [reactormonk/scala-elm-types#4]
Hi,

It's awesome that Options get converted to Maybes, this seems to work perfectly for primitives and case classes. Unfortunately the Codec fails with Options of List[X].

i.e. these work

```
object Working extends App {
  import shapeless.{::, HNil}
  import elmtype.ElmTypeShapeless._
  import elmtype._

  case class SomeType()
  case class OptionOfPrimitive(s: Option[String])
  case class OptionOfType(s: Option[SomeType])

  val types = ToElmTypes[OptionOfPrimitive :: OptionOfType :: HNil]().apply

  ElmTypeMain(types).main(List("client/Codec.elm").toArray)
}
```

but this does not:

```
object NotWorking extends App {
  import shapeless.{::, HNil}
  import elmtype.ElmTypeShapeless._
  import elmtype._

  case class OptionOfList(s: Option[List[String]])

  val types = ToElmTypes[OptionOfList :: HNil]().apply

  ElmTypeMain(types).main(List("client/Codec.elm").toArray)
}
```

The resulting Codec.elm fails to compile.

Be really useful if Option List was supported

p.

**Comment by alltonp** [comment]
Error is as follows:

```
==================================== ERRORS ====================================

-- TYPE MISMATCH --------------------------------------------------- ./Codec.elm

The 1st argument to function `maybe` is causing a mismatch.

10|                                                Decode.maybe Decode.list Decode.string)
                                                                ^^^^^^^^^^^
Function `maybe` is expecting the 1st argument to be:

    Decode.Decoder (List a)

But it is:

    Decode.Decoder a -> Decode.Decoder (List a)

Hint: It looks like a function needs 1 more argument.

-- TYPE MISMATCH --------------------------------------------------- ./Codec.elm

Function `maybe` is expecting 1 argument, but was given 2.

10|                                                Decode.maybe Decode.list Decode.string)
                                                                            ^^^^^^^^^^^^^
Maybe you forgot some parentheses? Or a comma?

-- TYPE MISMATCH --------------------------------------------------- ./Codec.elm

The right side of (<|) is causing a type mismatch.

13|                                            Maybe.map Encode.list <| List.map Encode.string obj.s)
                                                                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
(<|) is expecting the right side to be a:

    Maybe (List Encode.Value)

But the right side is:

    List Encode.Value

Hint: With operators like (<|) I always check the left side first. If it seems
fine, I assume it is correct and check the right side. So the problem may be in
how the left and right arguments interact.

-- TYPE MISMATCH --------------------------------------------------- ./Codec.elm

The 2nd argument to function `map` is causing a mismatch.

13|                                                                     List.map Encode.string obj.s)
                                                                                               ^^^^^
Function `map` is expecting the 2nd argument to be:

    List String

But it is:

    Maybe List String

Hint: I always figure out the type of arguments from left to right. If an
argument is acceptable when I check it, I assume it is "correct" in subsequent
checks. So the problem may actually be in how previous arguments interact with
the 2nd.

Detected errors in 1 module. 
```

**Comment by reactormonk** [comment]
I'd guess needs some more parens.

**Comment by reactormonk** [comment]
Yeah, but a bit more complicated than expected. I'd need a `$` similar to Haskell. The offending line: https://github.com/reactormonk/scala-elm-types/blob/master/core/src/main/scala/elmtype/derive/ast.scala#L112

**Comment by reactormonk** [comment]
Fixed in 0.4, on its way to maven central.
